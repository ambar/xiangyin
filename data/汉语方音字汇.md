# 汉语方音字汇

原书收录 20 个方言点的字音，湘语部分取长沙和双峰，长沙取新派口音。

## 数字化

[小学堂](https://xiaoxue.iis.sinica.edu.tw/)、[汉典](https://www.zdic.net/)、[古今文字集成](http://www.ccamc.co/)都收录了汉语方音字汇发音，目测它们的数据源相同，后两者都使用了小学堂的数据（都保留了同样的错误）并且没有收录多义字读音（只取了文白区别）。

湘语地区长沙和双峰的数据来自《汉语方音字汇》，全州和灌阳的数据来自《广西汉语方言研究》。

### 数据问题汇总

- 格式错误，详见[汉方字.简.修正.json](./raw/汉方字.简.修正.json)：
  - 许多字的数据列错位了，如「再」字：[汉典 - 再](https://www.zdic.net/zd/yy/xy/%E5%86%8D), [古今文字集成 - 再](http://www.ccamc.co/cjkv.php?cjkv=%E5%86%8D#xiang)
  - 错字：如「杯」中「阴平」写成音朋
  - 标点错误，声调标记错误等
- 交叉保存：例如同时收录了「云」与「雲」，全州/灌阳在一个字下，长沙/双峰在另一个字下
- 变体不当：例如沒有没有收录「間」，但把实际代表它（中间、间断）的字收录在「閒」(首义是 xián，次义是 jiān/jiàn）下，同时又收录了「閑」音

## 参考
- [小學堂湘語資料庫](https://xiaoxue.iis.sinica.edu.tw/xiangyu)
- [古音小鏡掃描檔](http://www.kaom.net/book_fangyinzihui.php)
