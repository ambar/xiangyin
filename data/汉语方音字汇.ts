import json from './raw/汉方字.简.修正.json'
import {csToneName2csToneNo, ToneName, ToneType} from './tones'
import {NormResult} from './types'

const schema = [
  ['號', String],
  ['字', String],
  [
    '湘',
    (v: string[][]) =>
      v.map((x) =>
        Object.fromEntries(x.map((x, i) => [subhead[i], x]))
      ) as unknown as RawDataSubItem,
  ],
] as const

const subhead = [
  '片',
  '方言点',
  '声母',
  '韵母',
  '调值',
  '调类',
  '备注',
] as const

type 县市 = '长沙' | '双峰' | '全州(县城)' | '灌阳(文市)'

type RawDataItem = {
  號: string
  字: string
  湘: RawDataSubItem[]
}

type RawDataSubItem = Record<typeof subhead[number], string>

type DataItem = {
  號: string
  字: string
  湘: Record<县市, NormResult[]>
}

const charGroup = new Map<string, DataItem[]>()

const parsePinyin = (v: RawDataSubItem): NormResult[] => {
  // 声/韵/调可能只在任意一个中出现斜线，主要是文白异读
  const i = v.声母.split('/')
  const f = v.韵母.split('/')
  const t = v.调类.split('/')
  const tv = v.调值.split('/')
  const max = Math.max(i.length, f.length, t.length)
  const fill = (list: string[], max: number) => {
    ;[...Array(max).keys()].forEach((_, i) => {
      if (!list[i]) list[i] = list[0]
    })
  }
  ;[i, f, t, tv].forEach((list) => fill(list, max))
  let note = ''
  let notes: string[] | void
  const def =
    max > 1
      ? v.备注
          .replace(/^(未注明异读原因)|(\p{Script=Han},[^.]+)/u, (x) => {
            note = x
            return ''
          })
          .replace(/^\p{Punctuation}/u, '')
      : v.备注
  if (note) {
    notes =
      note === '未注明异读原因' ? Array(5).fill('未注明') : note.split(/,/g)
  }
  return i.map((y, i) => {
    const tn = t[i]
    // 一些项目是单字表示，统一一下（入 -> 入声）
    const toneName = tn.length === 1 ? tn + '声' : tn
    return {
      音: y.replace('0', '') + f[i],
      调: toneName,
      释: (notes ? `〔${notes[i]}〕` : '') + def,
    }
  })
}

// TODO: 改成 script 预解析
export const items = json.map((x) => {
  const rawItem = Object.fromEntries(
    // @ts-expect-error
    schema.map(([key, val], i) => [key, val(x[i])])
  ) as unknown as RawDataItem
  const item = {
    號: rawItem.號,
    字: rawItem.字,
    湘: Object.fromEntries(rawItem.湘.map((v) => [v.方言点, parsePinyin(v)])),
  } as DataItem
  const setIfNotSet = (char: string) => {
    if (char) {
      if (charGroup.has(char)) charGroup.get(char)!.push(item)
      else charGroup.set(char, [item])
    }
  }
  setIfNotSet(item.字)
  return item
})

// @see 汉语方音字汇.md
const variantMap = [
  ['間', '閒'],
  // TODO: 应当更新 `queryVariants`（包含台湾用字），但要解决不上面的对应问题
  ['顏', '颜'],
]
variantMap.forEach(([a, b]) => {
  if (!charGroup.get(a) && charGroup.has(b)) {
    charGroup.set(a, charGroup.get(b)!)
  }
})

export const query = (char: string, cc: 县市, toneType?: ToneType) => {
  const items = charGroup.get(char)
  if (items && cc) {
    return items
      .map((x) => x.湘[cc] ?? [])
      .flat()
      .map((item) => {
        let tone = item.调
        if (toneType === 'CSToneNo') {
          tone = csToneName2csToneNo[tone as ToneName]!
        }
        return {...item, 调: tone}
      })
  }
  return []
}

export const queryCS = (char: string, toneType: ToneType = 'CSToneNo') => {
  return query(char, '长沙', toneType)
}
