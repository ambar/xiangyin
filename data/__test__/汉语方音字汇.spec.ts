import {queryCS} from '../汉语方音字汇'

const csChars = ['蟻', '指', '序', '樹', '核', '強', '會']
test.each(csChars)('queryCS: %s', (char) => {
  expect(queryCS(char)).toMatchSnapshot()
})

test('empty', () => {
  // 仅有灌阳/全州的结果
  expect(queryCS('鷺')).toEqual([])
  expect(queryCS('X')).toEqual([])
})
