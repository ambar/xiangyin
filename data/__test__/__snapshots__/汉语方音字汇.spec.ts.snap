// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`queryCS: 序 1`] = `
Array [
  Object {
    "调": 4,
    "释": "〔原〕",
    "音": "ɕy",
  },
  Object {
    "调": 4,
    "释": "〔新〕",
    "音": "ɕi",
  },
]
`;

exports[`queryCS: 指 1`] = `
Array [
  Object {
    "调": 6,
    "释": "〔未注明异读原因〕",
    "音": "tsɿ",
  },
  Object {
    "调": 3,
    "释": "〔未注明异读原因〕",
    "音": "tsɿ",
  },
]
`;

exports[`queryCS: 樹 1`] = `
Array [
  Object {
    "调": 5,
    "释": "〔文〕树立",
    "音": "ɕy",
  },
  Object {
    "调": 4,
    "释": "〔白〕树立",
    "音": "ɕy",
  },
  Object {
    "调": 5,
    "释": "〔文〕树林",
    "音": "ɕy",
  },
  Object {
    "调": 4,
    "释": "〔白〕树林",
    "音": "ɕy",
  },
]
`;

exports[`queryCS: 蟻 1`] = `
Array [
  Object {
    "调": 2,
    "释": "〔文〕",
    "音": "ȵi",
  },
  Object {
    "调": 5,
    "释": "〔白〕",
    "音": "ȵi",
  },
  Object {
    "调": 3,
    "释": "〔白〕",
    "音": "ȵi",
  },
  Object {
    "调": 4,
    "释": "〔白〕",
    "音": "ȵi",
  },
]
`;
